File: gametest/burnsale.py
#!/usr/bin/env python3

#   GSP for the Taurion blockchain game
#   Copyright (C) 2020  Autonomous Worlds Ltd
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""
Tests minting of vCHI through the burnsale.
"""

from pxtest import PXTest


class BurnsaleTest (PXTest):

  def run (self):
    self.collectPremine ()

    self.sendMove ("domob", {
      "vc": {"m": {}, "b": 10},
    }, burn=1.23456789)
    self.generate (1)

    a = self.getAccounts ()["domob"]
    self.assertEqual (a.getBalance (), 12345 - 10)
    self.assertEqual (a.data["minted"], 12345)

    self.sendMove ("domob", {"vc": {"m": {}}}, burn=2000000-1.2345)
    self.generate (1)

    a = self.getAccounts ()["domob"]
    self.assertEqual (a.getBalance (), 15000000000 - 10)
    self.assertEqual (a.data["minted"], 15000000000)

    self.assertEqual (self.getRpc ("getmoneysupply"), {
      "total": 15000000000,
      "entries":
        {
          "gifted": 0,
          "burnsale": 15000000000,
        },
      "burnsale":
        [
          {
            "stage": 1,
            "price": 0.0001,
            "total": 10000000000,
            "sold": 10000000000,
            "available": 0,
          },
          {
            "stage": 2,
            "price": 0.0002,
            "total": 10000000000,
            "sold": 5000000000,
            "available": 5000000000,
          },
          {
            "stage": 3,
            "price": 0.0005,
            "total": 10000000000,
            "sold": 0,
            "available": 10000000000,
          },
          {
            "stage": 4,
            "price": 0.0010,
            "total": 20000000000,
            "sold": 0,
            "available": 20000000000,
          },
        ],
    })


if __name__ == "__main__":
  BurnsaleTest ().main ()

File: src/burnsale.cpp
/*
    GSP for the Taurion blockchain game
    Copyright (C) 2020  Autonomous Worlds Ltd

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include "burnsale.hpp"

#include <glog/logging.h>

namespace pxd
{

Amount
ComputeBurnsaleAmount (Amount& burntChi, Amount alreadySold,
                       const Context& ctx)
{
  CHECK_GE (burntChi, 0);
  CHECK_GE (alreadySold, 0);

  Amount res = 0;
  for (const auto& stage : ctx.RoConfig ()->params ().burnsale_stages ())
    {
      /* Reduce the available amount in this stage by whatever was already
         sold previously (and has not yet been matched against previous
         stages in earlier iterations).  */
      Amount available = stage.amount_sold ();
      const Amount before = std::min (alreadySold, available);
      available -= before;
      alreadySold -= before;
      if (available == 0)
        continue;

      CHECK_GT (available, 0);
      CHECK_EQ (alreadySold, 0);

      /* Figure out how much of the remaining coins in this stage can
         be bought based on the price and available (burnt) CHI.  */
      const Amount affordable = burntChi / stage.price_sat ();
      const Amount sold = std::min (affordable, available);
      res += sold;
      burntChi -= sold * stage.price_sat ();

      /* If not all from this stage was bought, we are done.  Else continue
         trying future stages.  */
      if (sold < available)
        break;
      CHECK_EQ (sold, available);
    }

  return res;
}

} // namespace pxd
